Ash Drake Combat Buff

Whoneedspacee opened the pull request:
:cl: Whoneedspacee
add: new arena attack where ash drake summons lava around you
del: removed old swooping above you, instead flies above you instantly
balance: ash drake now spawns temporary lava pools instead of meteors falling down
balance: ash drake takes twice as long to swoop down now that he instantly goes above you
balance: ash drake now moves twice as fast
balance: increases the odds of lava spawns in the lava pool attack
balance: increases fire line damage and decreases lava attacks direct damage
tweak: ash drake fire now shoots in the direction of the target
tweak: changes times of certain animations
tweak: changes sounds of meteor falling to lava creation
fix: a bug where ash drakes attacks did not damage mechs
imageadd: changes meteor icon to lava creation animation from lava staff
/:cl:

[why]: Ash drake is one of the easiest megafauna despite giving fairly better loot than others like legion. Also he's a fairly boring boss to fight since he spends most of his time in the air with attacks that can't really hurt you ever as long as you keep moving. I did this to make his fight more fun and less one dimensional.

before - https://youtu.be/SjEJ5A_3-pI?t=14m50s

after - https://www.youtube.com/watch?v=jqJM3xySxBg&feature=youtu.be

Whoneedspacee committed:
Drake meteors are replaced with lava spawning out of the ground, deals damage on contact as well as normal lava damage, turns back after a period of time.

Fixes a bug where none of ash drakes attacks could damage mechs.

Adds directional fire to ash drake

Ash drake swoops no longer fly towards you but teleport above you, this was a non-threatening boring attack previously which is why it has been removed.

Slightly delays the time for meteors to hit the ground.

Adds cooldowns for the swoop attacks, previously there could be barely any time before ash drake would once again swoop up, unable to be attacked.

Whoneedspacee committed:
Ash drake now shoots fire when he cannot swoop.

Whoneedspacee committed:
Removes drakes ability to shoot fire after swoops, decreases swooping cooldown for triple swooping

Whoneedspacee committed:
Removes useless code and adds footsteps.

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/drake.dm:
```diff
@@ -104,70 +105,105 @@ Difficulty: Medium
+	var/turf/center = get_turf(target)
+	var/list/walled = RANGE_TURFS(3, center) - RANGE_TURFS(2, center)
+	for(var/turf/T in walled)
+		var/summoned_object_type = /obj/effect/forcefield/wizard
```
change the forcefield's icon to fire or something.

XDTM commented on code/modules/mob/living/simple_animal/hostile/megafauna/drake.dm:
```diff
@@ -263,71 +282,66 @@ Difficulty: Medium
+		var/reset_turf = T.type
+		T.ChangeTurf(lava_turf)
+		sleep(reset_time)
+		T.ChangeTurf(reset_turf)
```
PlaceOnTop might be better if you intend to reset the turf later

Whoneedspacee replied:
Not sure how I’d get rid of the turf on top afterwards, It works how I want it to though so I don’t really feel the need to change it unless it actually might be more efficient.

XDTM replied:
It will cause issues if for example a drake casts lava on already lava-ed turfs, which means that they'll remember the lava as the last turf instead of the original turf, making it permanent. Even if one drake alone can't do that, it might happen if two drakes are close to each other.

Whoneedspacee replied:
That's already in the code and it checks if the type if lava so that bug doesn't happen. There was a previous attack that could layer so that's why I made it.

XDTM commented on code/modules/mob/living/simple_animal/hostile/megafauna/drake.dm:
```diff
@@ -263,71 +282,66 @@ Difficulty: Medium
-	. = ..()
-	animate(src, pixel_z = 0, time = duration)
-
 /obj/effect/temp_visual/target
```
you should rename this to lava_warning or something like that, target is a bit too generic

Whoneedspacee replied:
Changing this to that actually did bring up the issue of the admin punishment using the drake effect for a fireball, should I move the old drake fireball code to the admin punishment file or just keep it in the drake file?

XDTM commented on code/modules/mob/living/simple_animal/hostile/megafauna/drake.dm:
```diff
@@ -104,70 +105,105 @@ Difficulty: Medium
+	swoop_attack(TRUE, null, 20) // longer cooldown on triple swoop ending
 
-/mob/living/simple_animal/hostile/megafauna/dragon/proc/swoop_attack(fire_rain, atom/movable/manual_target, swoop_duration = 40)
+/mob/living/simple_animal/hostile/megafauna/dragon/proc/swoop_attack(lava_pool = TRUE, atom/movable/manual_target, var/cooldown = 20)
```
the cooldown is in deciseconds, currently you've set it to 2 seconds

Whoneedspacee replied:
I do want it at 2 seconds, 20 seconds would be a bit excessive.

Whoneedspacee replied:
> it's for player drakes, and 2 seconds is absolutely bonkers.

It's not for player drakes, player drakes have their own seperate cooldown timer now.

XDTM commented on code/modules/mob/living/simple_animal/hostile/megafauna/drake.dm:
```diff
@@ -104,70 +105,105 @@ Difficulty: Medium
+	target.visible_message("<span class='boldwarning'>Lava starts to pool up beneath you!</span>")
+	for(var/turf/turf in range(10,get_turf(target)))
+		if(prob(35))
 			new /obj/effect/temp_visual/target(turf)
```
This is 441 checks, resulting in ~130 turf change operations on average, each of which has an icon change and tries to ignite every item on top, making them process the fire damage. This is expensive on its own, and worse if it's spammable by the drake.

Whoneedspacee replied:
I might rework this attack altogether because any less of a range and the player can just completely run away from the attack through the turning lava.

Whoneedspacee replied:
Does this new attack solve the problem?

MrDoomBringer commented:
"god fucking fire extinguishers why do you suck so much penis now"

had me loling

Barhandar commented:
Do you even fucking _play_ what the fuck is this "to fight drake you now must have equipment you best get from drake" bullshit.
> tweak: ash drake fire now shoots in the direction of the target

No. Just fucking no the whole point is to punish for staying at cardinals, SS13 is a fairly laggy multiplayer game, not singleplayer one where you can dodge all willy-nilly.
>I did this to make his fight more fun

This shit that gets you dead by single tiniest fucking lagspike because of all the lava is _NOT FUN_
>despite giving fairly better loot than others like legion.

Then buff the legion's loot instead of obliterating chances to kill the drake without getting HECK or lavaproof potion or mandatory Firefighter, which also drops more or less nothing valuable.
> fix: a bug where ash drakes attacks did not damage mechs

_All_ mechs or just the Firefighter? The latter is supposed to be immune to any fire damage.

Vidiveni commented:
The arena attack should spawn an extra clear spot for every extra person trapped inside of it.

epicmemename commented:
> Do you even fucking _play_ what the fuck is this "to fight drake you now must have equipment you best get from drake" bullshit.
In the videos that he posted, he showed himself that it's possible to kill it without ash drake gear or any megafauna gear. I don't know why you assume you would NEED gear since you can avoid every single attack fairly reliably. Joey boy even made swoops take twice as long as they would so that you wouldn't get killed instantly by them if you lagged

> No. Just fucking no the whole point is to punish for staying at cardinals, SS13 is a fairly laggy multiplayer game, not singleplayer one where you can dodge all willy-nilly.

Not staying at cardinals is hardly an engaging challenge. And you should still be fine even without lag since it allows you to go through the middle of it

> Then buff the legion's loot instead of obliterating chances to kill the drake without getting HECK or lavaproof potion or mandatory Firefighter, which also drops more or less nothing valuable.

You would die anyway if you got hit by a fireball before, since it could knock you out and then the drake would chomp you to death. This just makes it less RNG which is very needed, RNG in game mechanics is garbage. + lava you can survive, in fact every single hit you can survive with even 2 minutes of legion farming 

Joey boy said that he was going to rework every other megafauna anyway 
> _All_ mechs or just the Firefighter? The latter is supposed to be immune to any fire damage.

i think all mechs tbqhfam

ExcessiveUseOfCobblestone commented:
t. miner upset he won't get his free loot anymore.

subject217 commented:
Ash Drake is generally considered the easiest megafauna and these changes don't really make it impossible in any respect. This is a good PR IMO.

BebeYoshi commented:
Seems cool, however I have one question, does it makes the battle impossible for people that uses crusher (another don't use crusher git gud PR) or did you test it with crusher? (with and without watcher wing) Great job by the way

HydroFloric commented:
@BebeYoshi Seems doable with crusher but I mean drake is already pretty hard to do crusher wise because of his melee damage. Also good pr Joe.

Dorsisdwarf commented:
but i liked the swoop effect

Qustinnus commented:
keep the swoop tbf

Borisvanmemes commented:
What about lesser ash drakes, will this also buff them?

Whoneedspacee commented:
> What about lesser ash drakes, will this also buff them?

Stay tuned for a video in a bit, I'm showcasing all the changes I've made from this pr's recommendations.

ChangelingRain commented:
I feel like this is not the solution to the loot being better than a later/"stronger" boss, and you should probably rebalance the loot instead of buffing what's supposed-to-be the entry-level megafauna.

HydroFloric commented:
@ChangelingRain Honestly while this does make ash drake slightly more difficult the original drake fight was a fucking slog of him swooping for 50 years while you wait to actually be able to shoot them.

sentientloaf commented:
If this is because the loot of Ash Drake is to good why not nerf it?Ash Drake should be a sort of training megafauna so that non-miner mains can learn how to fight them.

Whoneedspacee committed:
Adds the old drake fireball into the admin punishments file so it can still be used

Removes the old mass lava changing attack, replaced with placing lava around you and swooping on top of that

Removes useless async calls

Adds a new wall to be used for the arena attack

Changes the name of the target visual to lava_warning for more clarity

Slows down the fire on the ash drake fire cone attack

There is now a spot for all mobs with clients in the arena attack

Whoneedspacee commented:
https://www.youtube.com/watch?v=jkOy4zvjdrQ&feature=youtu.be

Here are the changes I just commited in a video

Repukan commented:
Since when have megafauna been introduced with the concept of being entry level friendly? Where, in which PR, is this told or written?

The drake has been easy since its inception because nobody wanted to touch it (and any attempts to rebalance the loot were screeched out as well.) It was not easy because it was meant to be entry level friendly, it was easy because many changes made it easier to fight them as well. Mining and other department related changes were not tweaked in accordance to megafauna difficulty, because that was not the point of them, nor the point of balance changes, to keep megafauna difficult.

The whole point of having these megabeasts be the gate to powerful loot is that those that practice multiple times or those that luck out and beat them can reap the reward. To say and deny changes that make this harder because "new players/miners won't be able to enjoy them" is frankly fucking stupid and against the main idea of what they are, a minor challenge that is completely optional to those that wish to take it.

If you think it's hard, then don't fight them. But high resist armor, transformation spells, and even fucking magic items shouldn't be so easy to get that anyone with normal walking speed can do so. The drake is still beatable. You've been shown that those that want to, can even beat it without much preparation. These challenges and their rewards shouldn't be made easy just because of new player friendliness. There's no point to adding challenge and then turning around and saying,

DrChef commented:
I'm fine with megafauna being more difficult. Good job A+

Borisvanmemes commented:
The lava attack that makes you need to move to one spot seems...a tad hard to deal with from the videos, plus you always need to acount for lag, apart from that this seems like a good idea!

optimumtact commented:
none of you even know the beginning of the meaning of the word lag

daklaj commented:
rip killing drake with crusher
rip killing drake as ashwalkers

drderp3635 commented:
I would say to make the tiles about to be made into lava a tad bit more distinguishable from then when they are actually manifested

davethwave commented:
Don't complain about lag makes it impossible to defeat the ash drake now, if your having such bad lag then you shouldn't be mining at all unless you get scanned in cloning beforehand due to the likelihood of you falling into a chasm or lava lake. Also complaining about it in the first place is dumb, you can't kill any thing under the deadly power of lag, not even mentioning megafauna.

Whoneedspacee commented:
Last bug is the fire seems to damage himself despite the type check, then I’ll fix up the lava attack to be a bit more dangerous and that should be that.

ShizCalev commented:
Curious, how does the lava arena thing look if someone drags an ash drake onto the station?

Whoneedspacee commented:
I considered adding something for that, where it clears the inner arena of everything by the ash drake crushing it, because there isn’t much room on the station for that attack. Or I also considered if ash drake has no target on the station he flies back to lavaland, or being lazy and just leaving it like it is, what do you guys think?

ExcessiveUseOfCobblestone commented:
is leashing still a thing I haven't seen it happen in ages (t sybil player)

subject217 commented:
yes but it's very hard to do all the easy ways of doing it got removed

Whoneedspacee committed:
Removes useless swoop cooldown variable and uses parent megafauna mob recovery_time variable

Adds new lava swoop attack and a triplet variant to replace the old triple swooping attack

Replaces ash drake message warnings with src mob name for clarity

Fire line attack no longer includes source turf

Fixes a bug where the fire line attack would damage the ash drake himself

Fixes a bug where the ash drakes fire breath could hit you multiple times

Increases the speed of ash drakes fire attack

Alphas out the lava creation attacks more at the start of the animation for more discrepancy from already existing lava

Slightly increases the damage of direct lava hits to players

Whoneedspacee commented:
https://www.youtube.com/watch?v=ODJWhtFXmtc&feature=youtu.be

Besides these final changes there's only really one thing left to decide, should the arena attack clear the arena around it, or should the ash drake fly back to lavaland once it loses its target on the station for a period of time? Either of these would be appropriate for balancing the attack but I'd like to hear the reasons for one or the other, or another suggestion.

Boredone commented:
While not likely to come up that often, what about when multiple folks fight the Ash Drake at once, and it does the square trap thing while they are close to each other?
Is one of them just fucked currently or?

ChangelingRain reviewed (changes requested):
anyway you should definitely keep the swoop, and just make it move faster+last less time.

Togopalis commented:
Why would you want to keep the swoop? It was just an incredible time waster that added no difficulty. The arena that replaces it in this PR may not be the hardest, but easily is more challenging than "walk 1 tile away for each crosshair until he lands and hope he doesnt immediately take off again"

iskyp commented:
they want to keep the swoop because they made it look pretty @Togopalis

Whoneedspacee committed:
Changes turf in arena to basalt instead of being dumb and spawning the empty spaces in walls

Adds a waiting period in the lava arena attack so there's time to realize what attack is actually happening

Adds a new safe tile which makes it more apparent to the player where they need to go

Fixes a bug where the ash drake would shoot fire after dying in the middle of the lava swoop attack

Fixes a bug where the fire line attack could try to calculate target positioning without a target

Decreases fire breath direct damage

Re-adds swooping animation to fire drake but faster

Decreases direct lava damage again

Whoneedspacee commented:
@ChangelingRain https://www.youtube.com/watch?v=jqJM3xySxBg&feature=youtu.be

It's added, I've added everything I wanted to add, I'm not touching this anymore unless anyone else sees something else that's bad.

Barhandar commented:
>Why would you want to keep the swoop? It was just an incredible time waster that added no difficulty. 

Because the problem isn't with the swoop, it's with _it being able to be chained_. I.e. drakes needed a cooldown on using swoop.

Togopalis commented:
The replacement is still more difficult than it

Hathkar commented:
Oh no, I have to be awake to fight the drake now?

Whoneedspacee commented on code/modules/mob/living/simple_animal/hostile/megafauna/drake.dm:
```diff
@@ -201,33 +265,17 @@ Difficulty: Medium
-
+	
+	// Ash drake flies onto its target and rains fire down upon them
+	var/descentTime = 10;
```
Good point, it was balanced around it being a teleportation.

Whoneedspacee committed:
Slightly decreases mech direct damage

Removes useless swoop_speed variable

Decreases descent time variable

Removes useless goto_target() function

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into drake

Whoneedspacee committed:
Changes SetRecoveryTime method to work properly by changing the ranged cooldown, this makes it so setrecoverytime won't require manual checks in the openfire function, and will instead halt all instances of it being called for the time. This only affects the drake file as it's the only one that SetRecoveryTime is used in, but will aid in future megafauna alterations and cuts down on a check in the drake file.

Slightly increases the fire lines damage to be more threatening, it was face tankable previously.

Whoneedspacee commented:
100% not changing anything else, this was just something I noticed that could be cut down without affecting other megafauna currently and be helpful for other megafauna reworks later.

Barhandar commented:
You know, drake should be spawning chasms instead of lava, so he can't be cheesed with lavaproofing.

ExcessiveUseOfCobblestone commented:
so he can be cheesed by a much easier device that also allows the individual to escape!

XDTM commented:
Make it spawn supermatter

Whoneedspacee commented:
>           You know, drake should be spawning chasms instead of lava, so he can't be cheesed with lavaproofing.

He does direct damage to everything on spawns to counter this, so you can walk over lava after its spawned with a firefighter but not as it spawns, that will hit you. Same with the fire.

Barhandar commented:
>He does direct damage to everything on spawns to counter this, so you can walk over lava after its spawned with a firefighter but not as it spawns, that will hit you. Same with the fire.

I was intentionally trolling because your new drake is _unfightable, utter garbage_ and you've just confirmed that it's _intentional_ for the fight to be completely fucked.

Whoneedspacee commented:
How is it unfightable?

BHat0451 commented:
Ash Drake with this buff simply isn't as fun.

While Old Drake was admittedly a bit of a slog, it was still a newbie-friendly and manageable megafauna; which makes sense, considering it's likely the first true megafauna one would face. Cardinal-direction-only Fire Breath is easy to evade for more advanced players, but _potentially_ poses some troubles for more stubborn novice players. Fire Rain felt alright for a first-phase ability. The triple-swoop and fire-swoop felt somewhat underwhelming, but they were still more challenging than the mundane version. Swoop Speed felt okay; it meant that you could run from the fight if it got too challenging for you, and you could recoup your losses and return later, and Move Speed was slow, but acceptable; it worked.

New Drake is significantly more difficult, throwing the newbie-friendly part out the window. The lava-spawning effect, in my opinion, should be relegated to 'rage mode' at the very least, with the original fire-rain being pre-rage; that way, a person could get used to the fire raining down around them before it suddenly becomes lava-pools spawning around them when the drake reaches half health. Fire Rain during a swoop is a fair enough change, making the fight more exciting. Fire Breath could potentially be returned to its original, cardinal-direction-only form for pre-rage, before entering the New Drake breath for post-rage. Swoop Speed is utter bullshit, and move speed is almost just as bullshit. Old Swoop Speed works, and maybe a slightly-faster move speed than Old Drake would work; anything but New Drake move speed. While I have no gameplay experience with the 'arena' bit, just from watching it, it looks somewhat interesting; that is, if you don't lag at all. The moment you get FREE LAG in the arena, you would be burnt to a crisp; to say nothing of simple BYOND act-ups or what-not.

I personally think New Drake _could work_, but it just doesn't feel _right_ yet, from what little actual experience I have with it. Call me whatever, but New Drake just isn't as fun as it could be, in my opinion.

optimumtact commented:
new drake like new coke

HydroFloric commented:
@Daxxed Bubblegum is now the newb friendly boss.

ExcessiveUseOfCobblestone commented:
I heard that was getting messed with too so don't say that just yet lol
